src/org/opendatakit/briefcase/util/FormCache.java

```diff
@@ -1,97 +1,146 @@
-        UncheckedFiles.delete(cacheFile);
+        delete(cacheFilePath);
       }
-    UncheckedFiles.createFile(cacheFile);
-    return new FormCache(cacheFile, new HashMap<>(), new HashMap<>());
-  }
 
-  private void save() {
-    try (OutputStream out = Files.newOutputStream(cacheFile);
-         ObjectOutputStream oos = new ObjectOutputStream(out)) {
-      oos.writeObject(pathToMd5Map);
-      oos.writeObject(pathToDefinitionMap);
-    } catch (IOException e) {
-      throw new BriefcaseException("Can't serialize form cache", e);
+    // Check again since it could be deleted in the previous block
+    if (!Files.exists(cacheFilePath)) {
+      createFile(cacheFilePath);
+      hashByPath = new HashMap<>();
+      formDefByPath = new HashMap<>();
     }
   }
 
-  @Override
-  public String getFormFileMd5Hash(String filePath) {
-    return pathToMd5Map.get(filePath);
+  public void unsetLocation() {
+    briefcaseDir = Optional.empty();
+    cacheFile = Optional.empty();
+    hashByPath = new HashMap<>();
+    formDefByPath = new HashMap<>();
   }
 
-  @Override
-  public void putFormFileMd5Hash(String filePath, String md5Hash) {
-    pathToMd5Map.put(filePath, md5Hash);
+  private void save() {
+    cacheFile.ifPresent(path -> {
+      try (OutputStream out = Files.newOutputStream(path);
+           ObjectOutputStream oos = new ObjectOutputStream(out)) {
+        oos.writeObject(hashByPath);
+        oos.writeObject(formDefByPath);
+      } catch (IOException e) {
+        log.error("Can't serialize form cache", e);
+      }
+    });
   }
 
-  @Override
-  public BriefcaseFormDefinition getFormFileFormDefinition(String filePath) {
-    return pathToDefinitionMap.get(filePath);
+  public List<BriefcaseFormDefinition> getForms() {
+    return new ArrayList<>(formDefByPath.values());
   }
 
-  @Override
-  public void putFormFileFormDefinition(String filePath, BriefcaseFormDefinition definition) {
-    pathToDefinitionMap.put(filePath, definition);
+  public void update() {
+    briefcaseDir.ifPresent(path -> {
+      Set<String> scannedFiles = new HashSet<>();
+      list(path.resolve("forms")).forEach(formDir -> {
+        Path form = getFormFilePath(formDir);
+        scannedFiles.add(form.toString());
+        String hash = FileSystemUtils.getMd5Hash(form.toFile());
+        if (isFormNewOrChanged(form, hash)) {
+          try {
+            formDefByPath.put(form.toString(), new BriefcaseFormDefinition(form.getParent().toFile(), form.toFile()));
+            hashByPath.put(form.toString(), hash);
+          } catch (BadFormDefinition e) {
+            log.warn("Can't parse form file", e);
+          }
+        }
+      });
+      hashByPath.keySet().stream().filter(p -> !scannedFiles.contains(p)).collect(toList()).forEach(hashByPath::remove);
+      formDefByPath.keySet().stream().filter(p -> !scannedFiles.contains(p)).collect(toList()).forEach(formDefByPath::remove);
+      EventBus.publish(new CacheUpdateEvent());
```

Comment:
Do you need the collect here? Here’s my rewrite, with the common predicate extracted.
```
Predicate<String> notScanned = p -> !scannedFiles.contains(p);
hashByPath.keySet().stream().filter(notScanned).forEach(hashByPath::remove);
formDefByPath.keySet().stream().filter(notScanned).forEach(formDefByPath::remove);
```

Reply:
Is this a problem?
> If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.

Reply:
Yeah, the collect is to prevent that error. Otherwise, you are removing elements while you iterate the lists

Reply:
It's not very elegant. I'm open to suggetsions ;)

Reply:
Wow, look what IDEA just suggested.
```
hashByPath.keySet().removeIf(notScanned);
formDefByPath.keySet().removeIf(notScanned);
```
I didn’t check that it removes the elements from the map behind the key set view, but I think it does.

Reply:
That's neat! I'll add that

Reply:
I checked:
```
Map<String, String> m = new HashMap<>();
m.put("a", "aa");
m.put("b", "bb");
System.out.println(m);
m.keySet().removeIf(k -> k.equals("a"));
System.out.println(m);

{a=aa, b=bb}
{b=bb}
```

Reply:
Oh my, that's very troubling... maps are super mutable.